### PR TITLE
Add static types to iib/workers

### PR DESCRIPTION
**Pull request by xDaile**
CLOUDDST-14124

Static types to methods in ./iib/workers were added
Check in the tox file was configured also
Few type inconsistencies were repaired

### CODE REVIEW

File: iib/workers/config.py [hunk 1]
```diff
@@ -170,13 +172,14 @@ def configure_celery(celery_app):
...
         # a hack taken from flask.config.from_pyfile.
-        _user_config = {}
+        _user_config: Dict[str, Any] = {}
         with open(prod_config_file_path, mode='rb') as config_file:
             exec(compile(config_file.read(), prod_config_file_path, 'exec'), _user_config)  # nosec
 
         # Celery doesn't support configuring from multiple objects, so this is a way for
         # the configuration in prod_config_file_path to override the defaults in ProductionConfig
-        config = ProductionConfig()
+        # TODO is there some difference between ProductionConfig() and ProductionConfig - try debug
```
Comment (xDaile) [inline iib/workers/config.py:181]: I am not sure If this is okay, however tests passed

**Comment by xDaile** [comment]
@exd-hello-operator-guild PR is ready to be reviewed.

**Comment by yashvardhannanavati** [comment]
@release-engineering/exd-guild-hello-operator

File: iib/workers/api_utils.py [hunk 2]
```diff
@@ -5,13 +5,14 @@
 import requests
 from requests.packages.urllib3.util.retry import Retry
 import requests_kerberos
+from typing import Any, Dict, Optional
```
Comment (yashvardhannanavati) [inline iib/workers/api_utils.py:8]: looks like `typing` is a python standard library. This line should go after line 3 and before the empty line

File: iib/workers/config.py [hunk 3]
```diff
@@ -138,8 +140,8 @@ class DevelopmentConfig(Config):
         ],
     }
     iib_registry = 'registry:8443'
```
Comment (yashvardhannanavati) [inline iib/workers/config.py:142]: I see such strings not having static types. Is there a reason for that?

File: iib/workers/config.py [hunk 4]
```diff
@@ -170,12 +172,13 @@ def configure_celery(celery_app):
...
         # Celery doesn't support importing config files that aren't part of a Python path. This is
         # a hack taken from flask.config.from_pyfile.
-        _user_config = {}
+        _user_config: Dict[str, Any] = {}
         with open(prod_config_file_path, mode='rb') as config_file:
             exec(compile(config_file.read(), prod_config_file_path, 'exec'), _user_config)  # nosec
 
         # Celery doesn't support configuring from multiple objects, so this is a way for
         # the configuration in prod_config_file_path to override the defaults in ProductionConfig
+        # TODO is there some difference between ProductionConfig() and ProductionConfig - try debug
```
Comment (yashvardhannanavati) [inline iib/workers/config.py:181]: This line should probably be removed?

File: iib/workers/config.py [hunk 5]
```diff
@@ -256,7 +259,6 @@ def _validate_iib_org_customizations(iib_org_customizations):
         'image_name_from_labels': {'template'},
         'enclose_repo': {'enclosure_glue', 'namespace'},
     }
-
```
Comment (yashvardhannanavati) [inline iib/workers/config.py:259]: This is probably a mistake. Could you please undo this

File: iib/workers/greenwave.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 import logging
 
 import requests
+from typing import Any, Dict, List
```
Comment (yashvardhannanavati) [inline iib/workers/greenwave.py:7]: Same as above. This is a standard library

File: iib/workers/tasks/build_regenerate_bundle.py [hunk 7]
```diff
@@ -457,7 +484,7 @@ def _replace_csv_pullspecs(bundle_metadata, replacement_pullspecs):
     for old_pullspec, new_pullspec in replacement_pullspecs.items():
         log.debug(
             '%s will be replaced with %s in the bundle CSVs',
-            old_pullspec.to_str(),
+            old_pullspec.__repr__(),
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build_regenerate_bundle.py:487]: Could you please change this back to as it was. If you want this change to come in, could you please add it in a different PR

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]
Good work! This looks promising. A few comments

File: iib/workers/config.py [hunk 8]
```diff
@@ -10,31 +12,31 @@ class Config(object):
...
     iib_docker_config_template = os.path.join(
         os.path.expanduser('~'), '.docker', 'config.json.template'
     )
-    iib_greenwave_url = None
+    iib_greenwave_url: Optional[str] = None
     iib_grpc_init_wait_time = 30
     iib_grpc_max_port_tries = 100
     iib_grpc_max_tries = 5
     iib_grpc_start_port = 50051
     iib_image_push_template = '{registry}/iib-build:{request_id}'
```
Comment (JAVGan) [inline iib/workers/config.py:28]: I think these values are missing the static type as well as some other config values

File: iib/workers/tasks/build.py [hunk 9]
```diff
@@ -120,7 +121,11 @@ def _cleanup():
 
 
 @retry(exceptions=IIBError, tries=get_worker_config().iib_total_attempts, logger=log)
-def _create_and_push_manifest_list(request_id, arches, build_tags):
+def _create_and_push_manifest_list(
+    request_id: int,
+    arches: Iterable,
```
Comment (JAVGan) [inline iib/workers/tasks/build.py:126]: I believe this could be something like this:

```suggestion
    arches: Iterable[str],
```

File: iib/workers/tasks/build.py [hunk 10]
```diff
@@ -217,7 +222,7 @@ def _update_index_image_pull_spec(
     :raises IIBError: if the manifest list couldn't be created and pushed
     """
     conf = get_worker_config()
-    if from_index and overwrite_from_index:
+    if from_index and overwrite_from_index and resolved_prebuild_from_index:
```
Comment (JAVGan) [inline iib/workers/tasks/build.py:225]: Is this logic change required?

File: iib/workers/tasks/build.py [hunk 11]
```diff
@@ -641,8 +661,10 @@ def _skopeo_copy(source, destination, copy_all=False, exc_msg=None):
 
 
 def _verify_index_image(
-    resolved_prebuild_from_index, unresolved_from_index, overwrite_from_index_token=None
-):
+    resolved_prebuild_from_index: str,
+    unresolved_from_index: str,
+    overwrite_from_index_token: str = None,
```
Comment (JAVGan) [inline iib/workers/tasks/build.py:666]:
```suggestion
    overwrite_from_index_token: Optional[str] = None,
```

File: iib/workers/tasks/build_merge_index_image.py [hunk 12]
```diff
@@ -397,6 +397,8 @@ def is_bundle_version_valid(bundle_path, valid_ocp_version):
         raise IIBError(f'Invalid OCP version, "{valid_ocp_version}", specified in Index Image')
     try:
         bundle_version_label = get_image_label(bundle_path, 'com.redhat.openshift.versions')
+        if bundle_version_label is None:
+            raise ValueError
```
Comment (JAVGan) [inline iib/workers/tasks/build_merge_index_image.py:401]: I believe it would be better to raise an `IIBError` since it will be caught by API to properly log, also please consider including an error message for the error.

File: iib/workers/tasks/build.py (hunk 9 above)
Comment (xDaile) [reply iib/workers/tasks/build.py:126]: Sure, I missed that somehow.

File: iib/workers/tasks/build.py (hunk 11 above)
Comment (xDaile) [reply iib/workers/tasks/build.py:666]: Fixed

File: iib/workers/api_utils.py (hunk 2 above)
Comment (xDaile) [reply iib/workers/api_utils.py]: It should be okay in every file now.

File: iib/workers/tasks/build_merge_index_image.py (hunk 12 above)
Comment (xDaile) [reply iib/workers/tasks/build_merge_index_image.py]: Please check again. I made the change according to that.

File: iib/workers/tasks/build.py (hunk 10 above)
Comment (xDaile) [reply iib/workers/tasks/build.py]: My bad, did not noticed it is checked in functions called after that.

File: iib/workers/config.py (hunk 3 above)
Comment (xDaile) [reply iib/workers/config.py]: I intended to add static types just to method definitions, however I also added static types to class variables. Is it okay in this way?

File: iib/workers/config.py (hunk 8 above)
Comment (xDaile) [reply iib/workers/config.py]: I just added static types to all class variables (most of them in config.py). 
I intended to add static types just to method definitions at first, that´s why it was without them. 
Is it okay now?

Comment (JAVGan) [reply iib/workers/config.py]: Yes, thanks!

File: iib/workers/tasks/build_merge_index_image.py (hunk 12 above)
Comment (JAVGan) [reply iib/workers/tasks/build_merge_index_image.py]: Great, thanks!

File: iib/workers/config.py [hunk 13]
```diff
@@ -81,10 +83,10 @@ class ProductionConfig(Config):
...
     """The development IIB Celery configuration."""
 
-    broker_url = 'amqp://iib:iib@rabbitmq:5673//'
-    iib_api_url = 'http://iib-api:8080/api/v1/'
-    iib_log_level = 'DEBUG'
-    iib_organization_customizations = {
+    broker_url: str = 'amqp://iib:iib@rabbitmq:5673//'
+    iib_api_url: str = 'http://iib-api:8080/api/v1/'
+    iib_log_level: str = 'DEBUG'
+    iib_organization_customizations: Dict[str, Any] = {
```
[the remark this answers is not in the document]
Comment (xDaile) [reply iib/workers/config.py:89]: By this, you mean that iib_organization_customizations variable, or the DevelopmentConfig class?
By the way, thanks for the idea, I implemented TypeDict at different places.

File: iib/workers/tasks/build.py [hunk 14]
```diff
@@ -340,21 +350,22 @@ def _get_present_bundles(from_index, base_dir):
...
         unique_present_bundles.append(bundle)
         unique_present_bundles_pull_spec.append(bundle_path)
-
     return unique_present_bundles, unique_present_bundles_pull_spec
 
 
-def _get_missing_bundles(present_bundles, bundles):
+def _get_missing_bundles(
+    present_bundles: List[BundleImage],
+    resolved_bundles: List[str],
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:367]: Since the variable name changed, you probably need to change the docstring here too :)

File: iib/workers/tasks/build.py [hunk 15]
```diff
@@ -340,21 +350,22 @@ def _get_present_bundles(from_index, base_dir):
         return [], []
 
     # Transform returned data to parsable json
-    unique_present_bundles = []
+    unique_present_bundles: List[BundleImage] = []
     unique_present_bundles_pull_spec = []
     present_bundles = get_bundle_json(bundles)
-
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:346]: This was probably removed by accident? Could you please add the empty line back?

File: iib/workers/tasks/build.py [hunk 16]
```diff
@@ -340,21 +350,22 @@ def _get_present_bundles(from_index, base_dir):
...
     for bundle in present_bundles:
         bundle_path = bundle['bundlePath']
         if bundle_path in unique_present_bundles_pull_spec:
             continue
         unique_present_bundles.append(bundle)
         unique_present_bundles_pull_spec.append(bundle_path)
-
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:353]: Same as above

File: iib/workers/tasks/build.py [hunk 17]
```diff
@@ -485,12 +501,12 @@ def _opm_index_rm(base_dir, operators, binary_image, from_index, overwrite_from_
...
-    request_id,
-    output_pull_spec,
-    from_index,
-    resolved_prebuild_from_index,
-    overwrite_from_index_token=None,
-):
+    request_id: int,
+    output_pull_spec: str,
+    from_index: str,
+    resolved_prebuild_from_index: Optional[str],
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:507]: This probably should be `str`

File: iib/workers/tasks/build.py [hunk 18]
```diff
@@ -562,15 +581,15 @@ def _update_index_image_build_state(request_id, prebuild_info):
         image.
     """
     arches_str = ', '.join(sorted(prebuild_info['arches']))
-    payload = {
+    payload: Dict[str, Union[str, Dict[str, List[str]]]] = {
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:584]: Shouldn't this be `Dict[str, str]`?

File: iib/workers/tasks/build_merge_index_image.py [hunk 19]
```diff
@@ -125,7 +127,6 @@ def _add_bundles_missing_in_source(
         log.info(
             '%s bundles have invalid version label and will be deprecated.', len(invalid_bundles)
         )
-
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build_merge_index_image.py:128]: maybe indentation mistake? Could you please revert this?

File: iib/workers/tasks/build_regenerate_bundle.py [hunk 20]
```diff
@@ -287,11 +300,13 @@ def _adjust_operator_bundle(
             log.info('Resolving image pull specs')
             bundle_metadata = _get_bundle_metadata(operator_manifest, pinned_by_iib)
             _resolve_image_pull_specs(bundle_metadata, labels, pinned_by_iib)
-
     return labels
 
 
-def _get_bundle_metadata(operator_manifest, pinned_by_iib):
+def _get_bundle_metadata(
+    operator_manifest: OperatorManifest,
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/build_regenerate_bundle.py:307]: Nice! I like this

File: iib/workers/tasks/build_regenerate_bundle.py (hunk 7 above)
Comment (yashvardhannanavati) [reply iib/workers/tasks/build_regenerate_bundle.py:488]: I believe this is still pending? We discussed to remove __repr__ here

File: iib/workers/tasks/iib_types.py [hunk 21]
```diff
@@ -0,0 +1,51 @@
+from typing import TypedDict, Set, List, Dict, Optional
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/iib_types.py:1]: knitpick: Could we please rename this file to `iib_static_types`?

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: iib/workers/tasks/iib_types.py (hunk 21 above)
Comment (yashvardhannanavati) [reply iib/workers/tasks/iib_types.py:1]: Nice! I like this file

File: iib/workers/config.py [hunk 22]
```diff
@@ -2,6 +2,8 @@
 import os
 import logging
 import types
+from typing import Any, Dict, Optional, Type, Union, List
+from celery import Celery, app
```
[the remark this answers is not in the document]
Comment (xDaile) [reply iib/workers/config.py:6]: Yes, of course. That makes sense.

File: iib/workers/tasks/build.py (hunk 18 above)
Comment (xDaile) [reply iib/workers/tasks/build.py]: That´s right, thanks.

File: iib/workers/tasks/build.py (hunk 14 above)
Comment (xDaile) [reply iib/workers/tasks/build.py:369]: Thanks, I forgot to do that.

File: iib/workers/tasks/build_regenerate_bundle.py (hunk 7 above)
Comment (xDaile) [reply iib/workers/tasks/build_regenerate_bundle.py]: Should be okay now.

File: iib/workers/tasks/build.py (hunk 18 above)
Comment (xDaile) [reply iib/workers/tasks/build.py]: Actually, It was alright. 
The second Dict[str,List[str]] is needed for bundle_mapping assigned into the payload variable later (two lines after those selected).

Comment (xDaile) [reply iib/workers/tasks/build.py]: What is assigned into the payload['bundle_mapping'] in the tests looks like this:
`{'some-bundle': ['quay.io/some-bundle:v1'], 'some-bundle2': ['quay.io/some-bundle2:v1']}`

File: iib/workers/tasks/build.py [hunk 23]
```diff
@@ -120,7 +122,11 @@ def _cleanup():
 
 
 @retry(exceptions=IIBError, tries=get_worker_config().iib_total_attempts, logger=log)
-def _create_and_push_manifest_list(request_id, arches, build_tags):
+def _create_and_push_manifest_list(
+    request_id: int,
+    arches: Iterable[str],
```
Comment (chandwanitulsi) [inline iib/workers/tasks/build.py:127]: Why are we using Iterable here and not List?

File: iib/workers/tasks/build_merge_index_image.py [hunk 24]
```diff
@@ -153,31 +155,31 @@ def _add_bundles_missing_in_source(
     )
     _add_label_to_index(
         'com.redhat.index.delivery.distribution_scope',
-        distribution_scope,
+        distribution_scope,  # type: ignore
```
Comment (chandwanitulsi) [inline iib/workers/tasks/build_merge_index_image.py:158]: May I ask in what situations you have used type: ignore.
I noticed its used in function calls, local variables and some others

File: iib/workers/tasks/build.py (hunk 23 above)
Comment (xDaile) [reply iib/workers/tasks/build.py]: That´s a good point, I used that as i saw it elsewhere already used for arches. 
Now I changed arches, and all add_arches function parameters to have type Set - as the arches should not repeat itself. 
If I am wrong, please let me know, I will change those to List type.

File: iib/workers/tasks/build_merge_index_image.py (hunk 24 above)
Comment (xDaile) [reply iib/workers/tasks/build_merge_index_image.py:158]: Yes, of course. We discussed that with @yashvardhannanavati . We did not wanted to do big changes to logic of the IIB with this PR. 
Adding the types resulted in Mypy errors, which few were fixed, however few of them cannot be fixed without changing of the IIB logic (most of the times checking if the parameter is not None, I think this example is that case too).
So places where those errors were, are now annotated with those `# type: ignore` comments. After this PR is merged I will create another PR, where those comments will be removed, and fixes which are removing those errors proposed. 
Or I can make another commit which will fix those mypy errors and add it to this PR.

**Comment by xDaile** [comment]
Pipeline fail is not connected with changes made in this PR (docs build is failing).

File: iib/workers/tasks/build.py (hunk 23 above)
Comment (chandwanitulsi) [reply iib/workers/tasks/build.py]: Set is better :+1:

File: iib/workers/tasks/build_merge_index_image.py (hunk 24 above)
Comment (chandwanitulsi) [reply iib/workers/tasks/build_merge_index_image.py:158]: Another PR would be better :+1:

File: iib/workers/tasks/utils.py [hunk 25]
```diff
@@ -199,15 +209,21 @@ class RequestConfig:
...
         'cnr_token',
         'overwrite_from_index_token',
         'overwrite_target_index_token',
         'registry_auths',
     ]
 
-    _attrs = ["_binary_image", "distribution_scope", "binary_image_config"]
+    _attrs: List[str] = ["_binary_image", "distribution_scope", "binary_image_config"]
     __slots__ = _attrs
+    if TYPE_CHECKING:
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/utils.py:221]: I don't get this. Could you please add a comment explaining?

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]
Sorry for the last knitpick comments. But this is really taking shape now. Great job!

File: iib/workers/config.py [hunk 26]
```diff
@@ -81,87 +93,114 @@ class ProductionConfig(Config):
...
+                        'quay.io': 'registry.koji.company.com',
+                    },
+                }
+            ),
+            RelatedBundles({'type': 'related_bundles'}),
+            ImageNameFromLabels({'type': 'image_name_from_labels', 'template': '{name}-{version}'}),
+            EncloseRepo(
+                {'type': 'enclose_repo', 'enclosure_glue': '----', 'namespace': "company-pending"}
+            ),
+            RegistryReplacements(
```
[the remark this answers is not in the document]
Comment (xDaile) [reply iib/workers/config.py:171]: Only way, I found (Jindra advised me) that does not need breaking the datastructure into smaller parts.

File: iib/workers/tasks/iib_static_types.py [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+class BundleImage(TypedDict):
+    """Type class referencing data related to bundle image used for mypy checking."""
+
+    packageName: str
+    version: str
+    bundlePath: str
+    csvName: Optional[str]
+
+
+class ResolveImagePullspecs(TypedDict):
```
[the remark this answers is not in the document]
Comment (xDaile) [reply iib/workers/tasks/iib_static_types.py:55]: It could be also done without creating the class. 
Instead of class, there could be `resolve_image_pullspecs = TypedDict('resolve_image_pullspecs', {'type':str})`.

**Review by lipoja — CHANGES_REQUESTED** [review]

File: iib/workers/config.py [hunk 28]
```diff
@@ -2,76 +2,88 @@
...
+    iib_request_logs_dir: Optional[str] = None
+    iib_request_logs_format: str = (
         '%(asctime)s %(name)s %(processName)s {request_id} '
         '%(levelname)s %(module)s.%(funcName)s %(message)s'
     )
-    iib_request_logs_level = 'DEBUG'
-    iib_required_labels = {}
-    iib_request_related_bundles_dir = None
+    iib_request_logs_level: str = 'DEBUG'
+    iib_required_labels: Dict[str, Any] = {}
```
[the remark this answers is not in the document]
Comment (xDaile) [reply iib/workers/config.py:48]: Usually? Could you specify better the case when they are not `Dict[str, str]`? So I can change those parameters in right way.

File: iib/workers/tasks/iib_static_types.py [hunk 29]
```diff
@@ -0,0 +1,143 @@
...
+    bundle_image: Optional[str]
+    omps_operator_version: Optional[str]
+    target_index_resolved: Optional[str]
+    index_image_resolved: Optional[str]
+    internal_index_image_copy: Optional[str]
+    internal_index_image_copy_resolved: Optional[str]
+    index_image: Optional[str]
+    binary_image: Optional[str]
+    binary_image_resolved: Optional[str]
+    stage: Optional[str]
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/iib_static_types.py:79]: This should probably be removed.

File: iib/workers/tasks/build.py [hunk 30]
```diff
@@ -222,7 +228,7 @@ def _update_index_image_pull_spec(
             request_id,
             output_pull_spec,
             from_index,
-            resolved_prebuild_from_index,
+            resolved_prebuild_from_index,  # type: ignore
```
[the remark this answers is not in the document]
Comment (xDaile) [reply iib/workers/tasks/build.py:239]: Of course, Actually I added to the line above comment with the mypy error message - `            # MYPY error: Argument 4 to "_overwrite_from_index" has incompatible type "Optional[str]"; expected "str"`. There is a way, however I would like to create separated PR for those fixes.

File: iib/workers/api_utils.py [hunk 31]
```diff
@@ -156,7 +156,6 @@ def update_request(
         else:
             _exc_msg = f'The worker failed to update the request {request_id}'
         raise IIBError(_exc_msg)
-
```
Comment (yashvardhannanavati) [inline iib/workers/api_utils.py:152]: Probably a mistaken delete

File: iib/workers/tasks/iib_static_types.py [hunk 32]
```diff
@@ -52,49 +54,90 @@ class BundleImage(TypedDict):
     csvName: Optional[str]
 
 
-class ResolveImagePullspecs(TypedDict):
-    """TypedDict class for typing the DevelopmentConfig class."""
+class BinaryImageConfig(TypedDict):
+    """Type class referencing Binary Image configuration."""
 
-    type: str
+    prod: Dict[str, str]
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/iib_static_types.py:60]: I don't think `prod` is the right key name here. It's a string. Let's keep this Dict[str, Dict[str, str]]

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]
Let's squash these into a single commit. One final review and I think this should be +1 from me :)

File: iib/workers/tasks/iib_static_types.py (hunk 29 above)
Comment (xDaile) [reply iib/workers/tasks/iib_static_types.py]: You mean stage parameter?

File: iib/workers/tasks/iib_static_types.py (hunk 32 above)
Comment (xDaile) [reply iib/workers/tasks/iib_static_types.py]: I removed the class and replaced it with Dict[str, Dict[str, str]]. Pleas check.

File: iib/workers/tasks/build_merge_index_image.py [hunk 33]
```diff
@@ -153,31 +155,33 @@ def _add_bundles_missing_in_source(
     )
     _add_label_to_index(
         'com.redhat.index.delivery.distribution_scope',
-        distribution_scope,
+        # MYPY error: Argument 2 to "_add_label_to_index"
```
Comment (midnightercz) [inline iib/workers/tasks/build_merge_index_image.py:158]: In this case _add_label_to_index method should be updated to tolerate None as label or distribution scope should be changed make distribution_scope non-null

File: iib/workers/tasks/opm_operations.py [hunk 34]
```diff
@@ -171,10 +178,14 @@ def _serve_cmd_at_port(serve_cmd, cwd, port, max_tries, wait_time):
             ret = rpc_proc.poll()
             # process has terminated
             if ret is not None:
-                stderr = rpc_proc.stderr.read()
-                if 'address already in use' in stderr:
+                if rpc_proc.stderr is None:
```
Comment (chandwanitulsi) [inline iib/workers/tasks/opm_operations.py:181]: Can you please explain why we added failure when stderr is not captured?

Comment (xDaile) [reply iib/workers/tasks/opm_operations.py:181]: Of course. So this part of the code will raise Exception one way or other. However there was this MYPY error `error: Item "None" of "Optional[IO[str]]" has no attribute "read"`. Basically I just added check for value, that is used later (and can be possibly empty). 
I can make this change in separated PR, however I believe this is not change of logic - either way exception will be thrown.

File: iib/workers/tasks/build_merge_index_image.py (hunk 33 above)
Comment (xDaile) [reply iib/workers/tasks/build_merge_index_image.py:158]: Do I understand right that you are proposing those two solutions?
- None value should be accepted as the label_key parameter in the _add_label_to_index method
-  distribution_scope should be set to some empty value when it`s value is None
If I am right, which of that solutions is better according to you? Personally I like second one more, as it does not interfere with other part of the code - in some cases Mypy check that you are not sending None, could be useful.

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: iib/workers/tasks/opm_operations.py (hunk 34 above)
Comment (xDaile) [reply iib/workers/tasks/opm_operations.py]: Should be okay now.

Comment (xDaile) [reply iib/workers/tasks/opm_operations.py]: @chandwanitulsi stderr can be None, according to https://docs.python.org/3/library/subprocess.html#subprocess.CompletedProcess.stderr . I am not sure if it ever happened or if it is possible, however in case that `ret` would be set and `stderr` would be `None`, there would not be relevant error message. Please check solution now (error message was changed).

File: tests/test_workers/test_tasks/test_build_merge_index_image.py [hunk 35]
```diff
@@ -397,6 +397,7 @@ def test_add_bundles_missing_in_source(
         1,
         'amd64',
         '4.6',
+        'private',
```
Comment (yashvardhannanavati) [inline tests/test_workers/test_tasks/test_build_merge_index_image.py:400]: distribution_scope is usually `stage`, `prod` or `dev`

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]
Just one typo comment :)

File: tests/test_workers/test_tasks/test_build_merge_index_image.py (hunk 35 above)
Comment (xDaile) [reply tests/test_workers/test_tasks/test_build_merge_index_image.py]: 'Private' replaced for 'dev'. Is that okay?

File: iib/workers/tasks/iib_static_types.py [hunk 36]
```diff
@@ -0,0 +1,136 @@
...
+    target_index_resolved: str
+    target_ocp_version: str
+
+
+class BundleImage(TypedDict):
+    """Type class referencing data related to bundle image used for mypy checking."""
+
+    bundlePath: str
+    csvName: Optional[str]
+    packageName: str
```
Comment (JAVGan) [inline iib/workers/tasks/iib_static_types.py:53]: Don't we need `total=False` here as well since `csvName` can be optional?

File: iib/workers/tasks/iib_static_types.py [hunk 37]
```diff
@@ -0,0 +1,136 @@
...
+class AllIndexImagesInfo(TypedDict):
+    """Type class referencing group of IndexImageInfo classes used for mypy checking."""
+
+    from_index: IndexImageInfo
+    source_from_index: IndexImageInfo
+    target_index: IndexImageInfo
+
+
+# MYPY check - should label be optional? If not total parameter can be removed.
+class PrebuildInfo(TypedDict, total=False):
```
Comment (JAVGan) [inline iib/workers/tasks/iib_static_types.py:31]: AFAIK `labels` are indeed optional so it's fine to leave `total=False` there.

File: iib/workers/tasks/build_regenerate_bundle.py [hunk 38]
```diff
@@ -457,7 +486,7 @@ def _replace_csv_pullspecs(bundle_metadata, replacement_pullspecs):
     for old_pullspec, new_pullspec in replacement_pullspecs.items():
         log.debug(
             '%s will be replaced with %s in the bundle CSVs',
-            old_pullspec.to_str(),
+            old_pullspec,
```
Comment (JAVGan) [inline iib/workers/tasks/build_regenerate_bundle.py:489]: Why was this changed?

File: iib/workers/tasks/build_merge_index_image.py (hunk 24 above)
Comment (JAVGan) [inline iib/workers/tasks/build_merge_index_image.py:158]: It would be nice to put a comment before `_add_label_to_index` indicating the reason for this `type: ignore` like you did for the other ones.

**Review by JAVGan — COMMENTED** [review]
Just small comments.

Most of the work LGTM.

Also, please rebase this branch and solve the conflicts with `iib/workers/tasks/utils.py`

**Review by yashvardhannanavati — DISMISSED** [review]

File: iib/workers/tasks/iib_static_types.py (hunk 37 above)
Comment (xDaile) [reply iib/workers/tasks/iib_static_types.py:31]: Do we need those labels? It seems that they are passed to variable `prebuild_info` in the `handle_create_empty_index_request` method, however after that in the `_update_index_image_build_state` method they are not even used {I did not found any other place where those labels could be used}.

File: iib/workers/tasks/build_regenerate_bundle.py (hunk 38 above)
Comment (xDaile) [reply iib/workers/tasks/build_regenerate_bundle.py:489]: Type of the `old_pullspec` variable is already str.

File: iib/workers/tasks/build_merge_index_image.py (hunk 24 above)
Comment (xDaile) [reply iib/workers/tasks/build_merge_index_image.py:158]: Seems that it somehow slipped, this error was removed, with replacing of the `distribution_scope: Optional[str]` for `distribution_scope: str ` in the `_add_bundles_missing_in_source` method.

File: iib/workers/tasks/general.py [hunk 39]
```diff
@@ -13,7 +15,12 @@
 
 @app.task
 @request_logger
-def failed_request_callback(context, exc, traceback, request_id):
+def failed_request_callback(
+    context: celery.app.task.Context,
+    exc: Exception,
+    traceback,
```
[the remark this answers is not in the document]
Comment (xDaile) [reply iib/workers/tasks/general.py:21]: Mypy `reveal_type()` method is telling me that type is Any. I am not sure with the required type too, so I would propose leaving there type `Any`? If you can think of something better please let me know.

File: iib/workers/tasks/iib_static_types.py (hunk 37 above)
Comment (JAVGan) [reply iib/workers/tasks/iib_static_types.py]: The element `labels` is a supported optional parameter used only in `create-empty-index` requests, as you can see [here](https://iib.qe.engineering.redhat.com/#/paths/~1builds~1create-empty-index/post). It allows external entities to add additional labels to the resulting index image and if not specified we just set [these ones](https://github.com/release-engineering/iib/blob/master/iib/workers/tasks/build_create_empty_index.py#L126-L129).

If we remove the `labels` from IIB's code it will break this feature from its API on `create-empty-index` and will cause some troubles thus we need to keep it but consider this parameter to be optional from the requester point of view.

**Review by JAVGan — APPROVED** [review]
LGTM

File: iib/workers/tasks/iib_static_types.py (hunk 36 above)
Comment (xDaile) [reply iib/workers/tasks/iib_static_types.py]: Sure, updated. Thanks. However Optional[] does not mean that class has to have total=false, Optional means that parameter could be set to None value, and total=false means that it is possible that some of the keys will be missing.

File: iib/workers/tasks/iib_static_types.py [hunk 40]
```diff
@@ -37,23 +39,33 @@ class PrebuildInfo(TypedDict, total=False):
...
     ocp_version: str
     source_from_index_resolved: str
     source_ocp_version: str
     target_index_resolved: str
     target_ocp_version: str
 
 
-class BundleImage(TypedDict):
-    """Type class referencing data related to bundle image used for mypy checking."""
+class PrebuildInfo(PrebuildInfoBase, total=False):
```
Comment (yashvardhannanavati) [inline iib/workers/tasks/iib_static_types.py:49]: Hm, I think we can still use `NotRequired` in Python 3.8 if we use `typing_extensions.TypedDict` instead of `typing.TypedDict`

https://peps.python.org/pep-0655/#usage-in-python-3-11

**Comment by lipoja** [comment]
@xDaile There is missing `typing_extensions` dependence in `requirements-test.txt`

**Review by yashvardhannanavati — APPROVED** [review]

**Review by lipoja — APPROVED** [review]
